File: src/core/grammar/render/layout.rb
require 'core/expr/code/eval'
require 'core/expr/code/env'
require 'core/schema/tools/print'
require 'core/schema/code/factory'
require 'core/system/utils/paths'
require 'core/system/library/schema'
require 'core/semantics/code/interpreter'

module Layout
  
  # render an object into a grammar, to create a parse tree
  module RenderGrammar
    include Interpreter::Dispatcher

    def render(pat)
      stream = @D[:stream]
      @stack ||= []
      pair = "#{pat.to_s}/#{stream.current}"  # have to use strings, for JavaScript!
      if !@stack.include?(pair) # avoid infinite loop on recursive grammars
        @stack << pair 
        val = dispatch_obj(:render, pat)
        @stack.pop
        val
      end
    end

    def render_Grammar(this)
      stream = @D[:stream]
      @stack = []
      @create_stack = []
      @need_pop = 0
      # ugly, should be higher up
      @root = stream.current
      if @slash_keywords
        @literals = Scan.collect_keywords(this)
      end
      @modelmap = {}
      out = render(this.start.arg)
      format = {}
      format[:lines] = 0
      format[:space] = false
      format[:indent] = 0
      #puts @modelmap
      combine out, format
    end

    def render_Call(this)
      render(this.rule.arg)
    end

    def render_Alt(this)
      stream = @D[:stream]
      if @avoid_optimization
        this.alts.find_first do |pat|
          dynamic_bind stream: stream.copy do
            render(pat)
          end
        end
      else
        if !this.extra_instance_data
          this.extra_instance_data = []
          scan_alts(this, this.extra_instance_data)
          #puts "ALTS#{this.extra_instance_data}"
        end
        this.extra_instance_data.find_first do |info|
          pred = info[0]
          if !pred || pred.call(stream.current, @localEnv)
            dynamic_bind stream: stream.copy do
              render(info[1])
            end
          end
        end
      end
    end

    def render_Sequence(this)
      items = true
      ok = this.elements.all? do |x|
        item = render(x)
        if item
          if item == true
            true
          else
            if items.is_a?(Array)
              items << item
            elsif items != true
              items = [items, item]
            else
              items = item
            end
          end
        end
      end
      items if ok
    end

    def render_Create(this)
      stream = @D[:stream]
      obj = stream.current
      #puts "#{' '.repeat(@depth)}[#{this.name}] #{obj}"
      if !obj.nil? && obj.schema_class.name == this.name
        stream.next
        @create_stack.pop(@need_pop)
        @need_pop = @success = 0
        @create_stack.push [this, obj]
        res = dynamic_bind stream: SingletonStream.new(obj) do
          render(this.arg)
        end
        if res
          @success += 1
        end
        @need_pop += 1
        #this is the place to put some kind of tag around the res....
        if @add_tags and res!=nil
          format = {}
          format[:lines] = 0
          format[:space] = false
          format[:indent] = 0
          res = "*[*debug id='#{obj.schema_class.name}#{obj._id}'*]*" + combine(res, format) + "*[*/debug*]*"
        end
        @modelmap[res] = obj
        res
      else
        nil
      end
    end
  
    def render_Field(this)
      stream = @D[:stream]
      obj = stream.current
      # handle special case of [[ field:"text" ]] in a grammar 
      if this.arg.Lit?
        if this.arg.value == obj[this.name]
          this.arg.value
        end
      else
        #puts "#{' '.repeat(@depth)}FIELD #{this.name}"
        if this.name == "_id"
          data = SingletonStream.new(obj._id)
        else
          fld = obj.schema_class.all_fields[this.name]
          raise "Unknown field #{obj.schema_class.name}.#{this.name}" if !fld
#          path = obj.path().field(this.name)
          if fld.many
            data = ManyStream.new(obj[this.name])
          else
            data = SingletonStream.new(obj[this.name])
          end
        end
        res = dynamic_bind stream: data do
          render(this.arg)
        end
        if @add_tags and res!=nil
          format = {}
          format[:lines] = 0
          format[:space] = false
          format[:indent] = 0
          res = "*[*debug id='#{obj.schema_class.name}#{obj._id}#{this.name}'*]*" + combine(res, format) + "*[*/debug*]*"
        end
        res
      end
    end
    
    def render_Value(this)
      stream = @D[:stream]
      obj = stream.current
      if !obj.nil?
        if !(obj.is_a?(String) || obj.is_a?(Fixnum)  || obj.is_a?(Float))
          raise "Data is not literal #{obj}"
        end
        case this.kind
        when "str"
          if obj.is_a?(String)
            output(obj.inspect)
          end
        when "sym"
          if obj.is_a?(String)
            if @slash_keywords && @literals.include?(obj) then
              output('\\' + obj)
              output(obj)
            else
             output(obj)
            end
          end
        when "int"
          if obj.is_a?(Fixnum)
            output(obj.to_s)
          end
        when "real"
          if obj.is_a?(Float)
            output(obj.to_s)
          end
        when "atom"
          if obj.is_a?(String)
            output(obj.inspect)
          else
            output(obj.to_s)
          end
        else
          raise "Unknown type #{this.kind}"
        end
      end
    end
  
    def render_Ref(this)
      stream = @D[:stream]
      obj = stream.current
      if !obj.nil?
        # TODO: this is complete cheating! we need to search the path
        key_field = Schema::class_key(obj.schema_class)
        output(obj[key_field.name()])
      end
    end
  
    def render_Lit(this)
      stream = @D[:stream]
      obj = stream.current
      #puts "#{' '.repeat(@depth)}Rendering #{this.value} (#{this.value.class}) (obj = #{obj}, #{obj.class})"
      output(this.value)
    end
    
    def render_Code(this)
      stream = @D[:stream]
      obj = stream.current
      if this.expr.EBinOp? and this.expr.op=="eql?"
        rhs = Eval.eval(this.expr.e2, env: Env::ObjEnv.new(obj, @localEnv))
        if rhs.is_a? Factory::MObject
          lhs = Eval.eval(this.expr.e1, env: Env::ObjEnv.new(obj, @localEnv))
          puts "blah"
          lhs.schema_class == rhs.schema_class
        else
          Eval.eval(this.expr, env: Env::ObjEnv.new(obj, @localEnv))
        end
      else
        Eval.eval(this.expr, env: Env::ObjEnv.new(obj, @localEnv))
      end
    end
  
    def render_Regular(this)
      stream = @D[:stream]
      if !this.many
        # optional
        render(this.arg) || true
      else
        if stream.size > 0 || this.optional
          oldEnv = @localEnv
          @localEnv = Env::HashEnv.new
          @localEnv['_size'] = stream.size
          s = []
          i = 0
          ok = true
          while ok && stream.size > 0
            @localEnv['_index'] = i
            @localEnv['_first'] = (i == 0)
            @localEnv['_last'] = (stream.size == 1)
            if i > 0 && this.sep
              v = render(this.sep)
              if v
                s << v
              else
                ok = false
              end
            end
            if ok
              pos = stream.size
              v = render(this.arg)
              if v
                s << v
                stream.next if stream.size == pos
                i = i + 1
              else
                ok = false
              end
            end
          end
          @localEnv = oldEnv
          s if ok && (stream.size == 0)
        end
      end
    end
    
    def render_NoSpace(this)
      this
    end
    
    def render_Indent(this)
      this
    end
    
    def render_Break(this)
      this
    end
  
    def output(v)
      v
    end
    
    def scan_alts(this, alts)
      this.alts.each do |pat|
        if pat.Alt?
          scan_alts(pat, infos)
        else
          pred = PredicateAnalysis.new.recurse(pat)
          puts "pred = #{pred}"
          alts << [pred, pat]
        end
      end
    end

    def combine(obj, format)
      if obj == true
        # nothing
        ""
      elsif obj.is_a?(Array)
#        if @modelmap[obj]
#          o = @modelmap[obj]
#          @out << "<div id='<#{o.schema_class.name}:#{o._id}>' style='font-size: 48pt; color: fuchsia'>"
#          obj.each {|x| combine x}
#          @out << "</div>"
#        else
          res = ""
          obj.each {|x| res = res + combine(x, format)}
          res
#        end
      elsif obj.is_a?(String)
        res = ""
        if format[:lines] > 0
          res = res + ("\n".repeat(format[:lines]))
          res = res + (" ".repeat(format[:indent]))
          format[:lines] = 0
        else
          (res = res + " ") if format[:space]
        end
        res = res + obj
        format[:space] = true
        res
      elsif obj.NoSpace?
        format[:space] = false
        ""
      elsif obj.Indent?
        format[:indent] = format[:indent] + 2 * obj.indent
        ""
      elsif obj.Break?
        format[:lines] = System.max(format[:lines], obj.lines)
        ""
      else
        raise "Unknown format #{obj}"
      end
    end  

  end
  
  class PredicateAnalysis
  
    def recurse(pat)
      send(pat.schema_class.name, pat)
    end
  
    def Call(this)
      recurse(this.rule.arg)
    end
  
    def Alt(this)
      if this.alts.all? {|alt| alt.Field? && alt.arg.Lit? }
        fields = this.alts.map {|alt| alt.name }
        name = fields[0]
        #puts "ALT lits!! #{fields}"
        if fields.all? {|x| x == name}
          symbols = this.alts.map {|alt| alt.arg.value }
          lambda{|obj, env| symbols.include?(obj[name]) }
        end
      end
    end
  
    def Sequence(this)
      memo = nil
      this.elements.each do |x|
        pred = recurse(x)
        if memo && pred
          lambda{|obj, env| memo.call(obj, env) && pred.call(obj, env) }
        else
          memo ||= pred
        end
      end
      memo
    end
  
    def Create(this)
      name = this.name
      pred = recurse(this.arg)
      if pred
        lambda{|obj, env| !obj.nil? && obj.schema_class.name == name && pred.call(obj, env)}
      else
        lambda{|obj, env| !obj.nil? && obj.schema_class.name == name}
      end
    end
  
    def Field(this)
      name = this.name
      # handle special case of [[ field:"text" ]] in a grammar 
      if this.arg.Lit?
        value = this.arg.value
        lambda{|obj, env| value == obj[name]}
      elsif this.name != "_id"
        pred = recurse(this.arg)
        if pred
          lambda{|obj, env| pred.call(obj[name], env)}
        end
      end
    end
    
    def Value(this)
    end
  
    def Ref(this)
      lambda{|obj, env| !obj.nil?}
    end
  
    def Lit(this)
    end
  
    def Code(this)
      if this.schema_class.defined_fields.map{|f|f.name}.include?("code") && this.code != ""
       # FIXME: this case is needed to parse bootstrap grammar
        code = this.code.gsub("=", "==").gsub(";", "&&").gsub("@", "self.")
        lambda{|obj, env| obj.instance_eval(code) }
      else
        interp = Eval::EvalExprC.new
        lambda do |obj, env| 
          interp.dynamic_bind env: Env::ObjEnv.new(obj, env) do
            interp.eval(this.expr)
          end
        end
      end
    end
  
    def Regular(this)
      if this.many && !this.optional
        lambda{|obj, env| obj.size > 0 }
      end
    end
    
    def NoSpace(this)
    end
    
    def Indent(this)
    end
    
    def Break(this)
    end
  end
  
  class SingletonStream
    def initialize(data, used = false)
      @data = data
      @used = used
    end
    def size
      @used ? 0 : 1
    end
    def current
      @used ? nil : @data
    end
    def next
      @used = true
    end
    def copy
      SingletonStream.new(@data, @used)
    end
  end
  
  class ManyStream
    def initialize(collection, index = 0)
      @collection = collection.is_a?(Array) ? collection : collection.values 
      @index = index
      if @collection.include?(false)
        raise "not an object!!"
      end
    end
    def size
      @collection.size - @index
    end
    def current
      (@index < @collection.size) && @collection[@index]
    end
    def next
      @index = @index + 1
    end
    def copy
      ManyStream.new(@collection, @index)
    end
  end
 
  class DisplayFormat
    include RenderGrammar

    def self.print(*args)
      Layout::DisplayFormat.new.print(*args)
    end

    def print(grammar, obj, output=$stdout, slash_keywords = true, add_tags=false)
#      interp = RenderGrammarC.new
      @slash_keywords = slash_keywords
      @avoid_optimization = true
      @out = output
      @add_tags = add_tags
      res = dynamic_bind stream: SingletonStream.new(obj) do
        render(grammar)
      end
      output << res
      output << "\n"
      res
    end
  end
end

File: src/js/core/grammar/render/layout.js
'use strict'

//// Layout ////

var cwd = process.cwd() + '/';
var Eval = require(cwd + "core/expr/code/eval.js");
var Env = require(cwd + "core/expr/code/env.js");
var Print = require(cwd + "core/schema/tools/print.js");
var Factory = require(cwd + "core/schema/code/factory.js");
var Schema = require(cwd + "core/system/library/schema.js");
var Interpreter = require(cwd + "core/semantics/code/interpreter.js");
var Enso = require(cwd + "enso.js");

var Layout;

function RenderGrammar(parent) {
  return class extends Enso.mix(parent, Interpreter.Dispatcher) {
    render(pat) {
      var self = this, stream, pair, val;
      stream = self.D$.get$("stream");
      self.stack$ = self.stack$ || [];
      pair = Enso.S(pat.to_s(), "/", stream.current());
      if (! self.stack$.include_P(pair)) {
        self.stack$.push(pair);
        if (self.indent$) {
          puts(" " * self.indent$ + pat);
          self.indent$ = self.indent$ + 1;
        }
        val = self.dispatch_obj("render", pat);
        if (self.indent$) {
          self.indent$ = self.indent$ - 1;
        }
        self.stack$.pop();
        return val;
      }
    };

    render_Grammar(this_V) {
      var self = this, stream, out, format;
      stream = self.D$.get$("stream");
      self.stack$ = [];
      self.create_stack$ = [];
      self.need_pop$ = 0;
      self.root$ = stream.current();
      if (self.slash_keywords$) {
        self.literals$ = Scan.collect_keywords(this_V);
      }
      self.modelmap$ = Enso.EMap.new();
      out = self.render(this_V.start().arg());
      format = Enso.EMap.new();
      format .set$("lines", 0);
      format .set$("space", false);
      format .set$("indent", 0);
      return self.combine(out, format);
    };

    render_Call(this_V) {
      var self = this;
      return self.render(this_V.rule().arg());
    };

    render_Alt(this_V) {
      var self = this, stream, pred;
      stream = self.D$.get$("stream");
      if (self.avoid_optimization$) {
        return this_V.alts().find_first(function(pat) {
          return self.dynamic_bind(function() {
            return self.render(pat);
          }, Enso.EMap.new({stream: stream.copy()}));
        });
      } else {
        if (! this_V.extra_instance_data()) {
          this_V.set_extra_instance_data([]);
          self.scan_alts(this_V, this_V.extra_instance_data());
        }
        return this_V.extra_instance_data().find_first(function(info) {
          pred = info.get$(0);
          if (! pred || pred(stream.current(), self.localEnv$)) {
            return self.dynamic_bind(function() {
              return self.render(info.get$(1));
            }, Enso.EMap.new({stream: stream.copy()}));
          }
        });
      }
    };

    render_Sequence(this_V) {
      var self = this, items, item, ok;
      items = true;
      ok = this_V.elements().all_P(function(x) {
        item = self.render(x);
        if (item) {
          if (item == true) {
            return true;
          } else if (Enso.System.test_type(items, Array)) {
            return items.push(item);
          } else if (items != true) {
            return items = [items, item];
          } else {
            return items = item;
          }
        }
      });
      if (ok) {
        return items;
      }
    };

    render_Create(this_V) {
      var self = this, stream, obj, res, format;
      stream = self.D$.get$("stream");
      obj = stream.current();
      if (! (obj == null) && obj.schema_class().name() == this_V.name()) {
        stream.next();
        self.create_stack$.pop(self.need_pop$);
        self.need_pop$ = self.success$ = 0;
        self.create_stack$.push([this_V, obj]);
        res = self.dynamic_bind(function() {
          return self.render(this_V.arg());
        }, Enso.EMap.new({stream: SingletonStream.new(obj)}));
        if (res) {
          self.success$ = self.success$ + 1;
        }
        self.need_pop$ = self.need_pop$ + 1;
        if (self.add_tags$ && res != null) {
          format = Enso.EMap.new();
          format .set$("lines", 0);
          format .set$("space", false);
          format .set$("indent", 0);
          res = (Enso.S("*[*debug id='", obj.schema_class().name(), obj.identity(), "'*]*") + self.combine(res, format)) + "*[*/debug*]*";
        }
        self.modelmap$ .set$(res, obj);
        return res;
      } else {
        return null;
      }
    };

    render_Field(this_V) {
      var self = this, stream, obj, data, fld, res, format;
      stream = self.D$.get$("stream");
      obj = stream.current();
      if (Enso.System.test_type(this_V.arg(), "Lit")) {
        if (this_V.arg().value() == obj.get$(this_V.name())) {
          return this_V.arg().value();
        }
      } else {
        if (this_V.name() == "identity") {
          data = SingletonStream.new(obj.identity());
        } else {
          fld = obj.schema_class().all_fields().get$(this_V.name());
          if (! fld) {
            self.raise(Enso.S("Unknown field ", obj.schema_class().name(), ".", this_V.name()));
          }
          if (fld.many()) {
            data = ManyStream.new(obj.get$(this_V.name()));
          } else {
            data = SingletonStream.new(obj.get$(this_V.name()));
          }
        }
        res = self.dynamic_bind(function() {
          return self.render(this_V.arg());
        }, Enso.EMap.new({stream: data}));
        if (self.add_tags$ && res != null) {
          format = Enso.EMap.new();
          format .set$("lines", 0);
          format .set$("space", false);
          format .set$("indent", 0);
          res = (Enso.S("*[*debug id='", obj.schema_class().name(), obj.identity(), this_V.name(), "'*]*") + self.combine(res, format)) + "*[*/debug*]*";
        }
        return res;
      }
    };

    render_Value(this_V) {
      var self = this, stream, obj;
      stream = self.D$.get$("stream");
      obj = stream.current();
      if (! (obj == null)) {
        if (! ((Enso.System.test_type(obj, String) || Enso.System.test_type(obj, Enso.Numeric)) || Enso.System.test_type(obj, Float))) {
          self.raise(Enso.S("Data is not literal ", obj));
        }
        switch (this_V.kind()) {
          case "str":
            if (Enso.System.test_type(obj, String)) {
              return self.output(("\"" + obj) + "\"");
            }
          case "sym":
            if (Enso.System.test_type(obj, String)) {
              if (self.slash_keywords$ && self.literals$.include_P(obj)) {
                return self.output("\\" + obj);
              } else {
                return self.output(obj);
              }
            } else {
              return self.raise(Enso.S("Symbol is not a strign ", obj));
            }
          case "int":
            if (Enso.System.test_type(obj, Enso.Numeric)) {
              return self.output(obj.to_s());
            }
          case "real":
            if (Enso.System.test_type(obj, Float)) {
              return self.output(obj.to_s());
            }
          case "atom":
            if (Enso.System.test_type(obj, String)) {
              return self.output(("\"" + obj) + "\"");
            } else {
              return self.output(obj.to_s());
            }
          default:
            return self.raise(Enso.S("Unknown type ", this_V.kind()));
        }
      }
    };

    render_Ref(this_V) {
      var self = this, stream, obj, key_field;
      stream = self.D$.get$("stream");
      obj = stream.current();
      if (! (obj == null)) {
        key_field = obj.schema_class().key();
        return self.output(obj.get$(key_field.name()));
      }
    };

    render_Lit(this_V) {
      var self = this, stream, obj;
      stream = self.D$.get$("stream");
      obj = stream.current();
      return self.output(this_V.value());
    };

    render_Code(this_V) {
      var self = this, stream, obj, rhs, lhs;
      stream = self.D$.get$("stream");
      obj = stream.current();
      if (Enso.System.test_type(this_V.expr(), "EBinOp") && this_V.expr().op() == "eql?") {
        rhs = Eval.eval_M(this_V.expr().e2(), Enso.EMap.new({env: Env.ObjEnv.new(obj, self.localEnv$)}));
        if (Enso.System.test_type(rhs, Factory.MObject)) {
          lhs = Eval.eval_M(this_V.expr().e1(), Enso.EMap.new({env: Env.ObjEnv.new(obj, self.localEnv$)}));
          return lhs.schema_class() == rhs.schema_class();
        } else {
          return Eval.eval_M(this_V.expr(), Enso.EMap.new({env: Env.ObjEnv.new(obj, self.localEnv$)}));
        }
      } else {
        return Eval.eval_M(this_V.expr(), Enso.EMap.new({env: Env.ObjEnv.new(obj, self.localEnv$)}));
      }
    };

    render_Regular(this_V) {
      var self = this, stream, oldEnv, s, i, ok, v, pos;
      stream = self.D$.get$("stream");
      if (! this_V.many()) {
        return self.render(this_V.arg()) || true;
      } else if (stream.size_M() > 0 || this_V.optional()) {
        oldEnv = self.localEnv$;
        self.localEnv$ = Env.HashEnv.new();
        s = [];
        i = 0;
        ok = true;
        while (ok && stream.size_M() > 0) {
          self.localEnv$ .set$("_index", i);
          self.localEnv$ .set$("_first", i == 0);
          self.localEnv$ .set$("_last", stream.size_M() == 1);
          if (i > 0 && this_V.sep()) {
            v = self.render(this_V.sep());
            if (v) {
              s.push(v);
            } else {
              ok = false;
            }
          }
          if (ok) {
            pos = stream.size_M();
            v = self.render(this_V.arg());
            if (v) {
              s.push(v);
              if (stream.size_M() == pos) {
                stream.next();
              }
              i = i + 1;
            } else {
              ok = false;
            }
          }
        }
        self.localEnv$ = oldEnv;
        if (ok && stream.size_M() == 0) {
          return s;
        }
      }
    };

    render_NoSpace(this_V) {
      var self = this;
      return this_V;
    };

    render_Indent(this_V) {
      var self = this;
      return this_V;
    };

    render_Break(this_V) {
      var self = this;
      return this_V;
    };

    output(v) {
      var self = this;
      return v;
    };

    scan_alts(this_V, alts) {
      var self = this, pred;
      return this_V.alts().each(function(pat) {
        if (Enso.System.test_type(pat, "Alt")) {
          return self.scan_alts(pat, self.infos());
        } else {
          pred = PredicateAnalysis.new().recurse(pat);
          puts(Enso.S("pred = ", pred));
          return alts.push([pred, pat]);
        }
      });
    };

    combine(obj, format) {
      var self = this, res;
      if (obj == true) {
        return "";
      } else if (obj == null) {
        return self.raise("GRAMMAR FAILED TO PRODUCE OUTPUT");
      } else if (Enso.System.test_type(obj, Array)) {
        res = "";
        obj.each(function(x) {
          return res = res + self.combine(x, format);
        });
        return res;
      } else if (Enso.System.test_type(obj, String)) {
        res = "";
        if (format.get$("lines") > 0) {
          res = res + "\n".repeat(format.get$("lines"));
          res = res + " ".repeat(format.get$("indent"));
          format .set$("lines", 0);
        } else if (format.get$("space")) {
          res = res + " ";
        }
        res = res + obj;
        format .set$("space", true);
        return res;
      } else {
        switch (obj.schema_class().name()) {
          case "NoSpace":
            format .set$("space", false);
            return "";
          case "Indent":
            format .set$("indent", format.get$("indent") + 2 * obj.indent());
            return "";
          case "Break":
            format .set$("lines", Enso.System.max(format.get$("lines"), obj.lines()));
            return "";
          default:
            return self.raise(Enso.S("Unknown format ", obj));
        }
      }
    }; }};

class PredicateAnalysis {
  static new(...args) { return new PredicateAnalysis(...args) };

  recurse(pat) {
    var self = this;
    return self.send(pat.schema_class().name(), pat);
  };

  Call(this_V) {
    var self = this;
    return self.recurse(this_V.rule().arg());
  };

  Alt(this_V) {
    var self = this, fields, name, symbols;
    if (this_V.alts().all_P(function(alt) {
      return Enso.System.test_type(alt, "Field") && Enso.System.test_type(alt.arg(), "Lit");
    })) {
      fields = this_V.alts().map(function(alt) {
        return alt.name();
      });
      name = fields.get$(0);
      if (fields.all_P(function(x) {
        return x == name;
      })) {
        symbols = this_V.alts().map(function(alt) {
          return alt.arg().value();
        });
        return self.lambda(function(obj, env) {
          return symbols.include_P(obj.get$(name));
        });
      }
    }
  };

  Sequence(this_V) {
    var self = this, memo, pred;
    memo = null;
    this_V.elements().each(function(x) {
      pred = self.recurse(x);
      if (memo && pred) {
        return self.lambda(function(obj, env) {
          return memo(obj, env) && pred(obj, env);
        });
      } else {
        return memo = memo || pred;
      }
    });
    return memo;
  };

  Create(this_V) {
    var self = this, name, pred;
    name = this_V.name();
    pred = self.recurse(this_V.arg());
    if (pred) {
      return self.lambda(function(obj, env) {
        return (! (obj == null) && obj.schema_class().name() == name) && pred(obj, env);
      });
    } else {
      return self.lambda(function(obj, env) {
        return ! (obj == null) && obj.schema_class().name() == name;
      });
    }
  };

  Field(this_V) {
    var self = this, name, value, pred;
    name = this_V.name();
    if (Enso.System.test_type(this_V.arg(), "Lit")) {
      value = this_V.arg().value();
      return self.lambda(function(obj, env) {
        return value == obj.get$(name);
      });
    } else if (this_V.name() != "identity") {
      pred = self.recurse(this_V.arg());
      if (pred) {
        return self.lambda(function(obj, env) {
          return pred(obj.get$(name), env);
        });
      }
    }
  };

  Value(this_V) {
    var self = this;
  };

  Ref(this_V) {
    var self = this;
    return self.lambda(function(obj, env) {
      return ! (obj == null);
    });
  };

  Lit(this_V) {
    var self = this;
  };

  Code(this_V) {
    var self = this, code, interp;
    if (this_V.schema_class().defined_fields().map(function(f) {
      return f.name();
    }).include_P("code") && this_V.code() != "") {
      code = this_V.code().gsub("=", "==").gsub(";", "&&").gsub("@", "self.");
      return self.lambda(function(obj, env) {
        return obj.instance_eval(code);
      });
    } else {
      interp = Eval.EvalExprC.new();
      return self.lambda(function(obj, env) {
        return interp.dynamic_bind(function() {
          return interp.eval_M(this_V.expr());
        }, Enso.EMap.new({env: Env.ObjEnv.new(obj, env)}));
      });
    }
  };

  Regular(this_V) {
    var self = this;
    if (this_V.many() && ! this_V.optional()) {
      return self.lambda(function(obj, env) {
        return obj.size_M() > 0;
      });
    }
  };

  NoSpace(this_V) {
    var self = this;
  };

  Indent(this_V) {
    var self = this;
  };

  Break(this_V) {
    var self = this;
  };
};

class SingletonStream {
  static new(...args) { return new SingletonStream(...args) };

  constructor(data, used = false) {
    var self = this;
    self.data$ = data;
    self.used$ = used;
  };

  size_M() {
    var self = this;
    if (self.used$) {
      return 0;
    } else {
      return 1;
    }
  };

  current() {
    var self = this;
    if (self.used$) {
      return null;
    } else {
      return self.data$;
    }
  };

  next() {
    var self = this;
    return self.used$ = true;
  };

  copy() {
    var self = this;
    return SingletonStream.new(self.data$, self.used$);
  };
};

class ManyStream {
  static new(...args) { return new ManyStream(...args) };

  constructor(collection, index = 0) {
    var self = this;
    self.collection$ = Enso.System.test_type(collection, Array)
      ? collection
      : collection.values();
    self.index$ = index;
    if (self.collection$.include_P(false)) {
      self.raise("not an object!!");
    }
  };

  size_M() {
    var self = this;
    return self.collection$.size_M() - self.index$;
  };

  current() {
    var self = this;
    if (self.index$ < self.collection$.size_M()) {
      return self.collection$.get$(self.index$);
    } else {
      return null;
    }
  };

  next() {
    var self = this;
    return self.index$ = self.index$ + 1;
  };

  copy() {
    var self = this;
    return ManyStream.new(self.collection$, self.index$);
  };
};

class DisplayFormat extends Enso.mix(Enso.EnsoBaseClass, RenderGrammar) {
  static new(...args) { return new DisplayFormat(...args) };

  static print(grammar, obj, output = Enso.System.stdout(), slash_keywords = true, add_tags = false) {
    var self = this;
    return Layout.DisplayFormat.new().print(grammar, obj, output, slash_keywords, add_tags);
  };

  print(grammar, obj, output, slash_keywords, add_tags) {
    var self = this, res;
    self.slash_keywords$ = slash_keywords;
    self.avoid_optimization$ = true;
    self.out$ = output;
    self.add_tags$ = add_tags;
    res = self.dynamic_bind(function() {
      return self.render(grammar);
    }, Enso.EMap.new({stream: SingletonStream.new(obj)}));
    output.write(res);
    output.write("\n");
    return res;
  };
};

Layout = {
  RenderGrammar: RenderGrammar,
  PredicateAnalysis: PredicateAnalysis,
  SingletonStream: SingletonStream,
  ManyStream: ManyStream,
  DisplayFormat: DisplayFormat,
};
module.exports = Layout ;
